### PR TITLE
Feature/techdocs e2e

### LINKED ISSUE #5588: [TechDocs] Add a basic end-to-end test suite
## Feature Suggestion

Some basic functionality of TechDocs relies on interactions between the main Backstage context and context within the shadow DOM that holds the TechDocs site itself.  It's easy for changes in one area to break unrelated functionality.

We should cover some of these basic features and interactions with end-to-end tests, e.g. using cypress

## Possible Test Cases

- Navigation to and within a TechDocs site, including via the table of contents to the right, links within the doc, the primary navigation bar to the left, and directly from a URL, including hash/fragment links. [#5574]

(This list is not exhaustive. Add them as they come up or come to mind!)

### PR DESCRIPTION
**Pull request by ayshiff**
## Hey, I just made a Pull Request!

Closes #5588

This PR adds a basic e2e test-suite for TechDocs.
API responses (entities,`techdocs-backend`...) are mocked in the `fixtures` folder.

I have also: 

- added `login/logout` e2e tests (as those from the example app were failing)
- added some `data-testid` to some components to retrieve them more easily in the tests
- added a condition inside `plugins/techdocs/src/reader/components/TechDocsPage.tsx` if we try to fetch an entity that doesn't exist
- did some refactoring:
    - move all cypress commands inside `support/commands.ts`
    - renamed spec files into `**.spec.ts`
    - created a `API_BASE_URL` variable to be used as the API url and keep `http://localhost:3000` as the base front url inside the Cypress config

Here is the list of the different tests:

e2e-techdocs

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by freben** [comment]
So a thing that comes to mind immediately, is the large fixtures folder. Not in terms of size but in terms of maintenance - what happens as we evolve interfaces and API surfaces? For example, we're adding a `state` field to the entities response, and there are sure to be a lot of frontend changes as time goes on - what's the process to keep all of this fresh and updated even as kinda unrelated things are changed?

File: cypress/.eslintrc.js
Comment (freben) [inline cypress/.eslintrc.js:14]: Why is this needed?

File: plugins/techdocs/src/client.ts
Comment (freben) [inline plugins/techdocs/src/client.ts:101]: this should be above the call to .json above

**Review by freben — COMMENTED** [review]
just a few small comments, letting the techdocs team do the main reviewing

File: cypress/.eslintrc.js
Comment (ayshiff) [reply cypress/.eslintrc.js:14]: This is needed as BBD Assertions from Chai produce eslint errors. (https://docs.cypress.io/guides/references/assertions#BDD-Assertions)

Capture d’écran 2021-05-17 à 23 05 00

I wish I could find a cleaner solution.

**Comment by ayshiff** [comment]
> So a thing that comes to mind immediately, is the large fixtures folder. Not in terms of size but in terms of maintenance - what happens as we evolve interfaces and API surfaces? For example, we're adding a `state` field to the entities response, and there are sure to be a lot of frontend changes as time goes on - what's the process to keep all of this fresh and updated even as kinda unrelated things are changed?

I asked myself the same question before working on the issue.  
I wanted to separate frontend TechDocs tests from the API response as I thought that it is a good practice to separate the logic of both. 
But as you said, the frontend is strongly linked to the API response and I don't know how all this stuff will change in the future.

I think that the most maintainable solution in the long term is to not mock the API response and let the backend do its job. 
It would be nice to also find a way to add a `data-testid` attributes dynamically inside in the generated html files to make the tests easy to maintain.

I would love to have your opinion on this.

**Comment by ayshiff** [comment]
I have followed the advice of @freben and removed the TechDocs related fixtures and let the backend do its job. I only left the css mock because I had a MIME type error otherwise.

It should make the e2e tests easy to maintain for future changes.
I also added `intercept/wait` on the TechDocs related API calls to make sure everything is loaded before manipulating the DOM.

File: cypress/src/integration/plugins/techdocs.spec.ts
Comment (iamEAP) [inline cypress/src/integration/plugins/techdocs.spec.ts:81]: I kind of went back and forth in my head about the benefits/drawsbacks of relying on this documentation for the E2E tests, especially specific text on specific pages.

**For example**: It's possible for someone to make a change to this roadmap docs page in a branch...  And the E2E tests would pass (because it's loading docs from `master` on GitHub).  Once merged into into master, the E2E tests would start failing (because master was updated).

I wonder if it would be worth it to add a location for a separate fixture entity (in `tugboat.app-config.production.yaml`) whose `techdocs-ref` annotation used `dir:` instead of `url:`.  ...That way, the documentation used in E2E tests would always be loaded from the current branch, rather than what's in master.

...Of course the `dir` style of annotation is deprecated, but maybe we could save solving that for when we fully remove it.

What do you think?

File: cypress/src/integration/plugins/techdocs.spec.ts
Comment (iamEAP) [inline cypress/src/integration/plugins/techdocs.spec.ts:19]: While you're adding this, would you mind adding `@backstage/techdocs-core` as owner of the file in `.github/CODEOWNERS`?

**Review by iamEAP — COMMENTED** [review]
Thanks for the contribution @ayshiff!  A couple of quick thoughts on this (I'm only looking from a TechDocs perspective for the moment).

File: cypress/src/integration/plugins/techdocs.spec.ts
Comment (ayshiff) [reply cypress/src/integration/plugins/techdocs.spec.ts:81]: Yes, it makes sense to me too.

I don't know if this is what you meant in your comment but it would also be more relevant to make the e2e tests based on documentation that will not change in the future and that will be used only for these tests.

Comment (iamEAP) [reply cypress/src/integration/plugins/techdocs.spec.ts:81]: Yeah, adding a specific set of docs to act as test fixture sounds like a plan.

I was thinking it would also be beneficial for the docs to be read locally, rather than via URL, so that you could make changes to the fixture doc and the E2E tests in the same PR.  e.g.

```diff
# .tugboat/tugboat.app-config.production.yaml

backend:
  baseUrl: ${TUGBOAT_DEFAULT_SERVICE_URL}
  cors:
    origin: ${TUGBOAT_DEFAULT_SERVICE_URL}
    methods: [GET, POST, PUT, DELETE]
    credentials: true

+catalog:
+  locations:
+    - type: file
+      target: ../path/to/techdocs/e2e-fixture.catalog.info.yaml
```

With a corresponding info yaml like...

```yaml
# e2e-fixture.catalog.info.yaml

apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: techdocs-e2e-fixture
  description: Used for end-to-end tests of TechDocs in Backstage.
  annotations:
    backstage.io/techdocs-ref: dir:./
```

Where `backstage.io/techdocs-ref: dir:./` is used instead of `backstage.io/techdocs-ref: url: https://github.com/etc`.

Comment (ayshiff) [reply cypress/src/integration/plugins/techdocs.spec.ts:81]: Yeah, it sounds great!
I will make a custom set of docs to test the different TechDocs interactions.

A question that comes to my mind is about how to test this locally, outside of tugboat?
Should we add another command like `start-e2e-backend` that launch the backend with the correct set of docs?

Comment (iamEAP) [reply cypress/src/integration/plugins/techdocs.spec.ts:81]: Ah, actually [it looks like](https://discord.com/channels/687207715902193673/687207715902193679/846659061582659624) we're not even running tugboat builds automatically right now.  So maybe it's fine to just put it in `app-config.yaml` for now.

\cc @freben ^

Comment (Rugvip) [reply cypress/src/integration/plugins/techdocs.spec.ts:81]: Yeah tbh I don't see that adding those fixtures in the main `app-config.yaml` would hurt, so go ahead :grin: Especially compared to adding additional commands

**Comment by ayshiff** [comment]
I have added an e2e set of docs to run our Cypress tests against.

<p align="center">

</p>

I would like to have your opinion on a problem I have.

As I use lorem ipsum to simulate the content of the documentation, Vale complains that the quality of the document is not good.

For the local command I tried to filter the e2e files like for `OWNERS.md` and `ADOPTERS.md`.

```js
  const e2eFixturesDir = 'cypress/fixtures/docs/';
  const e2eFixtures = [
    'index.md',
    'sub-page-one.md',
    'sub-page-two.md',
    'sub-page-three.md',
  ].map(file => `${e2eFixturesDir}${file}`);

  const ignored = ['', 'ADOPTERS.md', 'OWNERS.md', ...e2eFixtures];
```

Or by prefixing the files with "e2e_".

```js
  const e2eFixtures = [
    'e2e_index.md',
    'e2e_sub-page-one.md',
    'e2e_sub-page-two.md',
    'e2e_sub-page-three.md',
  ]

 const ignored = ['', 'ADOPTERS.md', 'OWNERS.md', ...e2eFixtures];
```

The problem is that for the GitHub action, we don't have a way to specify the files we want to exclude. Knowing that `docs` is in the list of "files" we should specify all paths explicitly to avoid including test files.

```yaml
uses: errata-ai/vale-action@v1.3.0
with:
  files: '[".changeset", ".github", "contrib", "docs", "microsite", "packages", "plugins", "CONTRIBUTING.md", "CODE_OF_CONDUCT.md", "GOVERNANCE.md", "README.md"]'
```

I don't know what is the easiest way to avoid this problem

**Comment by Rugvip** [comment]
@ayshiff I'd say the easiest way is to just not use lorem ipsum. If you need to make scrolling or something else happen you could really just repeat some work over and over. Allow me to demonstrate :grin:

```md
To the margin! To the margin! To the margin! To the margin! To the margin! To the margin! To the margin! To the margin! To the margin! To the margin!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
Onwards to next page!
```

**Review by iamEAP — APPROVED** [review]
Thanks for the patience on this, @ayshiff!  Good w/this from a TechDocs POV.  No strong opinion on the vale lint quality issue you mentioned.

**Comment by ayshiff** [comment]
Sorry for the waiting time, I'm a little busy right now.
I'll look at this when I have a little more time!

File: cypress/src/support/commands.ts
Comment (benjdlambert) [inline cypress/src/support/commands.ts:21]: Does this still work? I'm pretty sure that we have to visit the page first to get the window object like before?

File: .changeset/cold-bags-admire.md
Comment (benjdlambert) [inline .changeset/cold-bags-admire.md:3]: This package doesn't exist anymore

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5675-ev5aw9pweqvltt9vusk61hlrshmd9gn0.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60f56c8780afb04f3944cd44

File: cypress/src/integration/integrations.spec.ts
Comment (benjdlambert) [inline cypress/src/integration/integrations.spec.ts:18]: we use `CYPRESS_baseUrl` as an override to set this so that we can run it against PR's. So I think we can revert this bit.

File: cypress/src/integration/user/login.spec.ts
Comment (benjdlambert) [inline cypress/src/integration/user/login.spec.ts:22]: This is not always present, maybe change it to `cy.contains('Select a sign-in method');`

File: packages/app/cypress/support/commands.js
Comment (benjdlambert) [inline packages/app/cypress/support/commands.js:69]: Don't think we can mock this out as it will change hash every time there's a new release. What's the purpose of mocking this stylesheet?

Comment (ayshiff) [reply packages/app/cypress/support/commands.js:69]: Yes you are right! I forgot to remove this mock when I removed the fixtures.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5675-ev5aw9pweqvltt9vusk61hlrshmd9gn0.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60f56c8780afb04f3944cd44

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5675-ev5aw9pweqvltt9vusk61hlrshmd9gn0.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60f56c8780afb04f3944cd44

**Review by Rugvip — APPROVED** [review]
:shipit: !

**Pull request merged**
